rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,234 @@
+- tarball: We can add `export TELEPORT_INSTALL_METHOD_TARBALL="true"` to the [`install`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/build.assets/install) script. (However, if the customer does not use the `install` script and instead moves the binaries manually, we won't be able to track this installation method.)
+- `.deb`/`.rpm`/`.pkg` packages, APT or YUM repository: It's unclear ATM how these can be tracked.
+- built from source: While it's technically possible for customers to build Teleport from source, we won't try to track this installation method as it seems an unlikely use-case.
+- `homebrew`: It's also possible to install Teleport on macOS using `homebrew`. The Teleport package in `homebrew` is not maintained by us, so we will also not track this installation method.
+
+In summary, for now we'll have the following new environment variables and respective value in `UpstreamInventoryHello.InstallMethods`:
+- `TELEPORT_INSTALL_METHOD_DOCKERFILE`: `dockerfile`
+- `TELEPORT_INSTALL_METHOD_HELM_KUBE_AGENT`: `helm-kube-agent`
+- `TELEPORT_INSTALL_METHOD_WINDOWS_SCRIPT`: `windows-script`
+- `TELEPORT_INSTALL_METHOD_NODE_SCRIPT`: `node-script`
+
+##### 8. Container runtime
+
+To determine if the agent is running on Docker, we'll check if the file `/.dockerenv` exists.
+Docker itself [does this](https://github.com/moby/libnetwork/blob/1f3b98be6833a93f254aa0f765ff55d407dfdd69/drivers/bridge/setup_bridgenetfiltering.go#L161).
+
+If we're interested in tracking other container runtimes, we could follow the approach by `gopsutil` ([here](https://github.com/shirou/gopsutil/blob/v3.23.1/internal/common/common_linux.go#L130-L278)).
+
+##### 9. Container orchestrator
+
+To determine if the agent is running on a Kubernetes pod, we can try to initialize a Kubernetes `client` similar to how [Validator.getClient()](https://github.com/gravitational/teleport/blob/master/lib/kubernetestoken/token_validator.go#L50-L68) does it.
+If this succeeds, the agent is running on Kubernetes.
+
+Afterwards, we'll try to detect in which cloud provider the pod is running on.
+For this, we'll call `client.ServerVersion()` and check if the returned `gitVersion` contains a certain substring specific for each provider:
+
+- in EKS, the git version looks like `"v1.24.8-eks-ffeb93d"` (we'll search for substring `"-eks"`)
+- in GPC ([docs](https://cloud.google.com/kubernetes-engine/docs/release-notes)), the git version looks like `"1.23.14-gke.1800"` (we'll search for substring `"-gke"`)
+
+In AKS, the git version looks like `"v1.25.2"`, so it's not possible to detect this environment using this method. (This is also a problem for Helm charts, as reported in [Azure/AKS#3375](https://github.com/Azure/AKS/issues/3375).)
+
+In the end, `UpstreamInventoryHello.ContainerOrchestrator` will be set to:
+- `kubernetes-eks` if on EKS
+- `kubernetes-gke` if on GKE
+- `kubernetes-unknown` otherwise (AKS, other cloud provider, or no cloud provider)
```

Comment:
Any concerns WRT slowing down start-up since all these checks will need to occur before the hello message sends?

Reply:
This is a good point.

It might be better to create a new message type instead of extending the `UpstreamInventoryHello`.
WDYT @fspmarshall?

Reply:
d9254a750dd0c01a7303466dbdc15b9c95e7c4f1 adds a new message type `UpstreamInventoryAgentMetadata` so that the existing ICS mechanism is not modified.